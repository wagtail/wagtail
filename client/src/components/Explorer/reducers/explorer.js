const defaultState = {
  isVisible: false,
  path: [],
};

/**
 * Oversees the state of the explorer. Defines:
 * - Where in the page tree the explorer is at.
 * - Whether the explorer is open or not.
 */
export default function explorer(prevState = defaultState, { type, payload }) {
  switch (type) {
  case 'OPEN_EXPLORER':
    // Provide a starting page when opening the explorer.
    return {
      isVisible: true,
      path: [payload.id],
    };

  case 'CLOSE_EXPLORER':
    return defaultState;

  case 'PUSH_PAGE':
    return {
      isVisible: prevState.isVisible,
      path: prevState.path.concat([payload.id]),
    };

  case 'POP_PAGE':
    return {
      isVisible: prevState.isVisible,
      path: prevState.path.slice(0, -1),
    };

  default:
    return prevState;
  }
}
