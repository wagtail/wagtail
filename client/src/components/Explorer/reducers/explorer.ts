export interface State {
  isVisible: boolean;
  path: number[];
}

const defaultState: State = {
  isVisible: false,
  path: [],
};

export const OPEN_EXPLORER = 'OPEN_EXPLORER';
interface OpenExplorerAction {
  type: typeof OPEN_EXPLORER;
  payload: {
    id: number;
  }
}

export const CLOSE_EXPLORER = 'CLOSE_EXPLORER';
interface CloseExplorerAction {
  type: typeof CLOSE_EXPLORER;
}

export const PUSH_PAGE = 'PUSH_PAGE';
interface PushPageAction {
  type: typeof PUSH_PAGE;
  payload: {
    id: number;
  }
}

export const POP_PAGE = 'POP_PAGE';
interface PopPageAction {
  type: typeof POP_PAGE;
}

export type Action = OpenExplorerAction | CloseExplorerAction | PushPageAction | PopPageAction;

/**
 * Oversees the state of the explorer. Defines:
 * - Where in the page tree the explorer is at.
 * - Whether the explorer is open or not.
 */
export default function explorer(prevState = defaultState, action: Action) {
  switch (action.type) {
  case OPEN_EXPLORER:
    // Provide a starting page when opening the explorer.
    return {
      isVisible: true,
      path: [action.payload.id],
    };

  case CLOSE_EXPLORER:
    return defaultState;

  case PUSH_PAGE:
    return {
      isVisible: prevState.isVisible,
      path: prevState.path.concat([action.payload.id]),
    };

  case POP_PAGE:
    return {
      isVisible: prevState.isVisible,
      path: prevState.path.slice(0, -1),
    };

  default:
    return prevState;
  }
}
