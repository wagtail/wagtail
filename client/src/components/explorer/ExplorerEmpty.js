import React from 'react';

const ExplorerEmpty = () => (
  <div className="c-explorer__placeholder">No results</div>
);

export default ExplorerEmpty;
