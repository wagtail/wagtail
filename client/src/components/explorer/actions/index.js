import { createAction } from 'redux-actions';

import { API_PAGES, PAGES_ROOT_ID } from 'config';

function _getHeaders() {
  const headers = new Headers();
  headers.append('Content-Type', 'application/json');

  return {
    credentials: 'same-origin',
    headers: headers,
    method: 'GET'
  };
}

function _get(url) {
  return fetch(url, _getHeaders()).then(response => response.json());
}

export const fetchStart = createAction('FETCH_START');

export const fetchSuccess = createAction('FETCH_SUCCESS', (id, body) => {
  return { id, body };
});

export const fetchFailure = createAction('FETCH_FAILURE');

export const pushPage = createAction('PUSH_PAGE');

export const popPage = createAction('POP_PAGE');

export const fetchBranchSuccess = createAction('FETCH_BRANCH_SUCCESS', (id, json) => {
  return { id, json };
});

export const fetchBranchStart = createAction('FETCH_BRANCH_START');

export const clearError = createAction('CLEAR_TRANSPORT_ERROR');

export const resetTree = createAction('RESET_TREE');

export const treeResolved = createAction('TREE_RESOLVED');

// Make this a bit better... hmm....
export function fetchTree(id = 1) {
  return (dispatch) => {
    dispatch(fetchBranchStart(id));

    return _get(`${API_PAGES}${id}/`)
      .then(json => {
        dispatch(fetchBranchSuccess(id, json));

        // Recursively walk up the tree to the root, to figure out how deep
        // in the tree we are.
        if (json.meta.parent) {
          dispatch(fetchTree(json.meta.parent.id));
        } else {
          dispatch(treeResolved());
        }
      });
  };
}

export function fetchRoot() {
  return (dispatch) => {
    // TODO Should not need an id.
    dispatch(resetTree(1));

    return _get(`${API_PAGES}?child_of=${PAGES_ROOT_ID}`)
      .then(json => {
        // TODO right now, only works for a single homepage.
        // TODO What do we do if there is no homepage?
        const rootId = json.items[0].id;

        dispatch(fetchTree(rootId));
      });
  };
}

export const toggleExplorer = createAction('TOGGLE_EXPLORER');

export const fetchChildrenSuccess = createAction('FETCH_CHILDREN_SUCCESS', (id, json) => {
  return { id, json };
});

export const fetchChildrenStart = createAction('FETCH_CHILDREN_START');

/**
 * Gets the children of a node from the API
 */
export function fetchChildren(id = 'root') {
  return (dispatch, getState) => {
    const { explorer } = getState();

    let api = `${API_PAGES}?child_of=${id}`;

    if (explorer.fields) {
      api += `&fields=${explorer.fields.map(global.encodeURIComponent).join(',')}`;
    }

    if (explorer.filter) {
      api = `${api}&${explorer.filter}`;
    }

    dispatch(fetchChildrenStart(id));

    return _get(api)
      .then(json => dispatch(fetchChildrenSuccess(id, json)));
  };
}

export function setFilter(filter) {
  return (dispatch, getState) => {
    const { explorer } = getState();
    const id = explorer.path[explorer.path.length - 1];

    dispatch({
      payload: {
        filter,
        id
      },
      type: 'SET_FILTER'
    });

    dispatch(fetchChildren(id));
  };
}

/**
 * TODO: determine if page is already loaded, don't load it again, just push.
 */
export function fetchPage(id = 1) {
  return dispatch => {
    dispatch(fetchStart(id));
    return _get(`${API_PAGES}${id}/`)
      .then(json => dispatch(fetchSuccess(id, json)))
      .then(json => dispatch(fetchChildren(id, json)))
      .catch(json => dispatch(fetchFailure(new Error(JSON.stringify(json)))));
  };
}

export const setDefaultPage = createAction('SET_DEFAULT_PAGE');
