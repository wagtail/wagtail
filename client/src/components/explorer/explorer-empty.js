import React from 'react';


const ExplorerEmpty = () => {
  return <div className='c-explorer__placeholder'>No results</div>
}

export default ExplorerEmpty;
