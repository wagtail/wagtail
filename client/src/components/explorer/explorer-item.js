import React, { Component, PropTypes } from 'react';
import StateIndicator from 'components/state-indicator';

export default class ExplorerItem extends Component {
  constructor(props) {
    super(props);
    this.state = {};
  }

  render() {
    const { title, data } = this.props;

    return (
      <div className="c-explorer__item">
        <h3 className="c-explorer__title">
          <StateIndicator state={data.state} />
          {title}
        </h3>
      </div>
    );
  }
}


ExplorerItem.propTypes = {
  title: PropTypes.string,
  data: PropTypes.object
};
