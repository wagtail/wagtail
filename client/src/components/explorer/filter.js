import React, { Component } from 'react';

const Filter = ({label, filter=null, activeFilter, onFilter}) => {
  let click = onFilter.bind(this, filter);
  let isActive =  activeFilter === filter;
  let cls = ['c-filter'];

  if (isActive) {
    cls.push('c-filter--active');
  }

  return (
    <span className={cls.join(' ')} onClick={click}>{label}</span>
  );
}


export default Filter;
