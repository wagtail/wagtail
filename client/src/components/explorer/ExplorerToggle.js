import React from 'react';
import { connect } from 'react-redux';

import * as actions from './actions';

import Button from '../../components/Button/Button';

/**
 * A Button which toggles the explorer, and doubles as a loading indicator.
 */
// TODO isVisible should not be used here, but at the moment there is a click
// binding problem between this and the ExplorerPanel clickOutside.
const ExplorerToggle = ({ isVisible, isFetching, children, onToggle }) => (
  <Button
    icon="folder-open-inverse"
    isLoading={isFetching}
    onClick={isVisible ? null : onToggle}
  >
    {children}
  </Button>
);

ExplorerToggle.propTypes = {
  isVisible: React.PropTypes.bool,
  isFetching: React.PropTypes.bool,
  onToggle: React.PropTypes.func,
  children: React.PropTypes.node,
};

const mapStateToProps = (store) => ({
  isFetching: store.explorer.isFetching,
  isVisible: store.explorer.isVisible,
});

const mapDispatchToProps = (dispatch) => ({
  onToggle() {
    dispatch(actions.toggleExplorer());
  },
});

export default connect(mapStateToProps, mapDispatchToProps)(ExplorerToggle);
