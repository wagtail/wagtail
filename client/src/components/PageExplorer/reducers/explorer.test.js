import explorer from './explorer';

describe('explorer', () => {
  const initialState = explorer(undefined, {});

  it('exists', () => {
    expect(explorer).toBeDefined();
  });

  it('returns the initial state if no input is provided', () =>  {
    expect(explorer(undefined, {})).toEqual(initialState);
  });

  it('OPEN_EXPLORER', () => {
    const action = { type: 'OPEN_EXPLORER', payload: { id: 1 } };
    expect(explorer(initialState, action)).toMatchSnapshot();
  });

  it('CLOSE_EXPLORER', () => {
    expect(explorer(initialState, { type: 'CLOSE_EXPLORER' })).toEqual(initialState);
  });

  it('PUSH_PAGE', () => {
    expect(explorer(initialState, { type: 'PUSH_PAGE', payload: { id: 100 } })).toMatchSnapshot();
  });

  it('POP_PAGE', () => {
    const state = explorer(initialState, { type: 'PUSH_PAGE', payload: { id: 100 } });
    const action = { type: 'POP_PAGE', payload: { id: 100 } };
    expect(explorer(state, action)).toMatchSnapshot();
  });
});
