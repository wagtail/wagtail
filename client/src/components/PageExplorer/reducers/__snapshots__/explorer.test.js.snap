// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`explorer OPEN_EXPLORER 1`] = `
Object {
  "currentPageId": 1,
  "depth": 0,
}
`;

exports[`explorer POP_PAGE 1`] = `
Object {
  "currentPageId": null,
  "depth": 0,
}
`;

exports[`explorer PUSH_PAGE 1`] = `
Object {
  "currentPageId": null,
  "depth": 0,
}
`;
